File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/handler/invoke/Invoker.java
package io.github.oleksivio.telegram.bot.core.controller.handler.invoke;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import io.github.oleksivio.telegram.bot.api.model.result.HandlerResult;
import io.github.oleksivio.telegram.bot.core.model.ITelegram;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public abstract class Invoker<ARG extends ITelegram> {
    private static final Logger LOG = LoggerFactory.getLogger(Invoker.class);

    private final Method method;
    private final Object classInstance;

    Invoker(Object classInstance,
            Method method) {
        this.method = method;
        this.classInstance = classInstance;
    }

    public HandlerResult run(ARG argument) {

        try {
            return invoke(classInstance, method, argument);
        }
        catch (IllegalAccessException | InvocationTargetException e) {
            RuntimeException invalidBehaviorException =
                    new IllegalStateException("Can't run ErrorResponse method " + method);
            LOG.error("", invalidBehaviorException);
            LOG.error("", e);
            throw invalidBehaviorException;

        }
        catch (ClassCastException ignored) {
            RuntimeException invalidBehaviorException =
                    new IllegalStateException("Invalid result type of method. "
                                                      + "HandlerResult or void was expected "
                                                      + method);
            LOG.error("", invalidBehaviorException);
            throw invalidBehaviorException;
        }

    }

    protected abstract HandlerResult invoke(Object classInstance,
                                            Method method,
                                            ARG argument) throws InvocationTargetException, IllegalAccessException;

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/handler/invoke/Invoker.kt
package io.github.oleksivio.telegram.bot.core.controller.handler.invoke

import io.github.oleksivio.telegram.bot.api.model.result.HandlerResult
import io.github.oleksivio.telegram.bot.core.model.ITelegram
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import kotlin.reflect.KFunction

abstract class Invoker<ARG : ITelegram> internal constructor(private val classInstance: Any,
                                                             private val func: KFunction<*>) {

    fun run(argument: ARG): HandlerResult {

        try {
            return invokeFunction(classInstance, classInstance.javaClass.getMethod(func.name), argument)
        } catch (e: IllegalAccessException) {
            val invalidBehaviorException = IllegalStateException("Can't run method $func")
            LOG.error("", e)
            throw invalidBehaviorException

        } catch (e: InvocationTargetException) {
            val invalidBehaviorException = IllegalStateException("Can't run method $func")
            LOG.error("", e)
            throw invalidBehaviorException
        } catch (ignored: ClassCastException) {
            val invalidBehaviorException = IllegalStateException("Invalid result type of method. "
                    + "HandlerResult or void was expected "
                    + func)
            LOG.error("", invalidBehaviorException)
            throw invalidBehaviorException
        }

    }

    @Throws(InvocationTargetException::class, IllegalAccessException::class)
    protected abstract fun invokeFunction(classInstance: Any,
                                          method: Method,
                                          argument: ARG): HandlerResult

    companion object {
        private val LOG = LoggerFactory.getLogger(Invoker::class.java)
    }

}
